### PR TITLE
Issue #7301: fix handling of small images

### PR DESCRIPTION
**Pull request by jo2**
Issue: #7301
* fix check for wrapper (checked only config_*.xml files til now)
* add handling of small images
* fix display of source

### CODE REVIEW

File: src/site/resources/css/site.css [hunk 1]
```diff
@@ -52,7 +52,6 @@ h2:hover .anchor, h3:hover .anchor {
 }
 
 .source {
-  width: 90%;
```
Comment (rnveach) [inline src/site/resources/css/site.css]: This was added in https://github.com/checkstyle/checkstyle/issues/3902 .

> https://romani.github.io/7301/site/writingjavadocchecks.html

I took one spot and zoomed in far to make the text span past the border.
Image1

It created a scrollbar on the bottom and could scroll the text, but the max text hit the border on the end.

https://github.com/checkstyle/checkstyle/pull/3998#issuecomment-287549880
Original change gave some padding between the border and the last text.

Can we add a right padding similar to the left?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java [hunk 2]
```diff
@@ -88,9 +88,19 @@ private static void iterateNode(Node node, String fileName, String sectionName)
                 Assert.assertTrue(wrapperMessage,
                         node.getAttributes().getNamedItem("class").getNodeValue()
                                 .contains("wrapper"));
+
                 if ("table".equals(child.getNodeName())) {
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java]: Sorry that this is on old code. Do we need to iterate the children of div/span/p or do we just not have them?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java [hunk 3]
```diff
@@ -88,9 +88,19 @@ private static void iterateNode(Node node, String fileName, String sectionName)
...
                     iterateNode(child, fileName, sectionName);
                 }
+                if ("img".equals(child.getNodeName())) {
+                    final String dataImageInlineMessage = fileName + "/" + sectionName + ": img "
+                            + "needs an attribute data-image-inline which has the value 'true' if "
+                            + "the image should be displayed in the same line as the surrounding "
+                            + "text.";
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java]: I don't understand the message versus the actual check.

Shouldn't the message say something like "node needs block or inline class attribute"?

File: src/xdocs/property_types.xml [hunk 4]
```diff
@@ -709,15 +709,19 @@
             <td><code>always</code></td>
             <td>
               Example:
-              <pre>@Deprecated()</pre>
+              <div class="wrapper">
+                <pre>@Deprecated()</pre>
```
Comment (rnveach) [inline src/xdocs/property_types.xml]: Why was this changed? This isn't an image or long text. Is this just because of the `pre`? If so, do we need a junit to verify the `pre`s too?

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java (hunk 2 above)
Comment (jo2) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java]: I don't exactly get want you mean there.. We have to itereate over every node because some p nodes contain imgs or pre elements, also spans and div.. we iterate them by using the iterateNode()-Method recursively.. I hope I could answer this..

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java (hunk 3 above)
Comment (jo2) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java]: Yeah, I will update this accordingly.

File: src/xdocs/property_types.xml (hunk 4 above)
Comment (jo2) [reply src/xdocs/property_types.xml]: yeah, this has to be wrapped, too. There was a bug in the unit test which is mentioned in the pr description, which causes only the config_*.xml files to be checked. this is one of the fixes which had to be done after fixing the test.

`pre` is a tag which is displayed in html, `source` gets rendered to a `pre` in a `div class="source"`. I don't know the exact difference and why sometimes a `source` and sometimes a `pre` is used tough.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java (hunk 2 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java]: > We have to itereate over every node because some p nodes contain imgs or pre elements

I see now, this is being done with `NODES_TO_WRAP`. The highlighted recursive call is only for `table` that also needs to be wrapped as well as children examined.

File: src/xdocs/property_types.xml (hunk 4 above)
Comment (rnveach) [reply src/xdocs/property_types.xml]: ok. Thanks.

File: src/site/resources/css/site.css (hunk 1 above)
Comment (jo2) [reply src/site/resources/css/site.css]: Sadly, that is not that easy because it is not possible to add css to text nodes (which the codefragments are). I can try more thorough tomorrow though.

Comment (rnveach) [reply src/site/resources/css/site.css]: @jo2 Why can't you add a padding to the `source` class listed above? It is the box that needs the padding.

Comment (romani) [reply src/site/resources/css/site.css]: @rnveach , can you explain me problem one more time.... I do not see what is wrong on big screen and on mobile for `Will lead to violation if` code block

**Comment by romani** [comment]
is there is nuances/problems with expanding source width, lets move it to separate PR, and merge fix for images.
it will simplify PR review, no need to fix all at ones.

File: src/site/resources/css/site.css (hunk 1 above)
Comment (romani) [reply src/site/resources/css/site.css]: hmmm, I do not see this as problem, and we can improve it later on, if there possibility to do this.
I am not sure we will benefit from extra wrapping of source to something else and make source without border and border of wrap.
Please think one more time about this point.
I am ok without gap.

Comment (rnveach) [reply src/site/resources/css/site.css]: I am not asking for extra wrapping, only extra padding. Text will stay on the same line as there is a scrollbar on the bottom. I am just hoping we can make some room between border and last of text.

I am fine with improving it later but I am still waiting for the message to be updated in the assertion.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java (hunk 2 above)
Comment (jo2) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java]: the recursive call iterates through every node in the dom and checks if it is one of `NODES_TO_WRAP`. If it is not or it is a `table`, the children are also checked, because DOM children (svg, pre, img excluded) can contain `NODES_TO_WRAP` elements.

File: src/site/resources/css/site.css (hunk 1 above)
Comment (jo2) [reply src/site/resources/css/site.css]: @rnveach There is a padding on the box. But the `pre` is overflowing to the right side so the padding on the right is ignored. The `pre` overflows til the border of the `source` because the `source` has `overflow: scroll` so everything which overflows the `source` will be scrolled. But we can swap the overflow from the `source` to the generated `pre` and add the padding there (see screenshot).

image

Comment (rnveach) [reply src/site/resources/css/site.css]: I am not sure if you are saying this can't be fixed. We can leave it for another time.

Comment (jo2) [reply src/site/resources/css/site.css]: this is a screenshot from my lokal build. I can upload it to my github pages if you want

Comment (rnveach) [reply src/site/resources/css/site.css]: Yes please upload. It is fixed in the current commit right? I just wanted to verify that is the changes I see now.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java (hunk 3 above)
Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsMobileWrapperTest.java]: This looks done and is better.

File: src/site/resources/css/site.css (hunk 1 above)
Comment (jo2) [reply src/site/resources/css/site.css]: Uploaded at https://jo2.github.io/checkstyle/

Comment (rnveach) [reply src/site/resources/css/site.css]: Thanks.

Comment (rnveach) [reply src/site/resources/css/site.css]: @jo2 On your version I am not seeing a scrollbar anymore or the padding on the right.

Comment (jo2) [reply src/site/resources/css/site.css]: It works in Chrome.. I'll take a look at it and try to make it work in firefox.

**Comment by romani** [comment]
IDEA volations are same as in master ... I am merging this PR.

File: src/site/resources/css/site.css (hunk 1 above)
Comment (jo2) [reply src/site/resources/css/site.css]: I now got a solution working in chrome and firefox. Uploaded it to https://jo2.github.io/checkstyle/

Comment (rnveach) [reply src/site/resources/css/site.css]: thanks @jo2 . It looks good to me. Feel free to submit another PR on it.
